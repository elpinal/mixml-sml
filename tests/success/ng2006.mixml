link X =
  link {module Tree =
    link T = {type t : Type} with
    {val max : ext T.t -> int}
  } with
  {module Forest =
    link T = {type t : Type} with
    {val max : ext T.t -> int}
  }
seals

link
{module Tree =
  link S = {type u = [Leaf : int, Node : int * ext X.Forest.t]} with
  link T = new [unit {data t = ext S.u}] with
  {val max = fun (x : ext T.t) ->
    match ext T.t_out x => {
      | Leaf i => i
      | Node(i, f) => ext X.Forest.max f
    }
  }
}
with

{module Forest =
  link S = {type u = [Nil : unit, Cons : ext X.Tree.t * ext X.Forest.t]} with
  link T = new [unit {data t = ext S.u}] with
  {val max = fun (x : ext T.t) ->
    match ext T.t_out x => {
      | Nil () => 0
      | Cons(hd, tl) => ext X.Tree.max hd
    }
  }
}
