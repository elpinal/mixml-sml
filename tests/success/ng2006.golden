{Forest : {max : [!0 -> int]+, t : [= !0]}, Tree : {max : [!1 -> int]+, t : [= !1]}}
